src/test/app/TxQ_test.cpp

```diff
@@ -2555,6 +2554,127 @@ class TxQ_test : public beast::unit_test::suite
+
+        auto const bobSeq = env.seq(bob);
+        // Ledger 4 gets 2 from bob and 1 from alice,
+        // Ledger 5 gets 4 from bob,
+        // Ledger 6 gets 5 from bob.
+        for (int i = 0; i < 2 + 4 + 5; ++i)
+        {
+            env(noop(bob), seq(bobSeq + i), fee(200), ter(terQUEUED));
+        }
+        checkMetrics(env, 10 + 2 + 4 + 5, boost::none, 2, 1, 256);
+        // Close ledger 3
+        env.close();
+        checkMetrics(env, 9 + 4 + 5, 20, 3, 2, 256);
+        BEAST_EXPECT(env.seq(alice) == aliceSeq + 12);
+        
+        // Close ledger 4
+        env.close();
+        checkMetrics(env, 9 + 5, 30, 4, 3, 256);
+        BEAST_EXPECT(env.seq(alice) == aliceSeq + 12);
+        
+        // Close ledger 5
+        env.close();
+        // Three of Alice's txs expired.
+        checkMetrics(env, 6, 40, 5, 4, 256);
+        BEAST_EXPECT(env.seq(alice) == aliceSeq + 12);
+
+        // Top off Alice's queue again using Tickets so the sequence gap is
+        // unaffected.
+        env(noop(alice), ticket::use(aliceSeq + 1), ter(terQUEUED));
+        env(noop(alice), ticket::use(aliceSeq + 2), ter(terQUEUED));
+        env(noop(alice), ticket::use(aliceSeq + 3), ter(terQUEUED));
+        env(noop(alice), ticket::use(aliceSeq + 4), ter(terQUEUED));
+        env(noop(alice), ticket::use(aliceSeq + 5), ter(terQUEUED));
+        env(noop(alice), ticket::use(aliceSeq + 6), ter(telCAN_NOT_QUEUE_FULL));
+        checkMetrics(env, 11, 40, 5, 4, 256);
+
+        // Even though alice's queue is full we can still slide in a couple
+        // more transactions because she has a sequence gap.  But we
+        // can only install a transaction that fills the bottom of the gap.
+        // Explore that...
+
+        // Verify that we can't queue a sequence-based transaction that
+        // follows a gap.
+        env(noop(alice), seq(aliceSeq + 20), ter(telCAN_NOT_QUEUE_FULL));
+
+        // Verify that the transaction in front of the gap is still present
+        // by attempting to replace it without a sufficient fee.
+        env(noop(alice), seq(aliceSeq + 15), ter(telCAN_NOT_QUEUE_FEE));
+
+        // We can't queue a transaction into the middle of the gap.  It must
+        // go at the front.
+        env(noop(alice), seq(aliceSeq + 18), ter(telCAN_NOT_QUEUE_FULL));
+        env(noop(alice), seq(aliceSeq + 17), ter(telCAN_NOT_QUEUE_FULL));
+
+        // Successfully put this transaction into the front of the gap.
+        env(noop(alice), seq(aliceSeq + 16), ter(terQUEUED));
+
+        // Still can't put a sequence-based transaction at the end of the gap.
+        env(noop(alice), seq(aliceSeq + 18), ter(telCAN_NOT_QUEUE_FULL));
+
+        // But we can still fill the gap from the front.
+        env(noop(alice), seq(aliceSeq + 17), ter(terQUEUED));
+
+        // Finally we can fill in the entire gap.
+        env(noop(alice), seq(aliceSeq + 18), ter(terQUEUED));
+        checkMetrics(env, 14, 40, 5, 4, 256);
```

Comment:
Add one more `env(noop(alice), seq(aliceSeq + 20), ter(telCAN_NOT_QUEUE_FULL));` after you've filled the gap to verify that you can't queue anything at the end now that the gap is full.